server-openid4vci/src/main/java/com/android/identity/server/openid4vci/openid4vp.kt

```diff
@@ -0,0 +1,233 @@
+import kotlinx.datetime.DateTimePeriod
+import kotlinx.datetime.TimeZone
+import kotlinx.datetime.plus
+import kotlinx.serialization.json.JsonObject
+import kotlinx.serialization.json.JsonPrimitive
+import kotlinx.serialization.json.add
+import kotlinx.serialization.json.buildJsonArray
+import kotlinx.serialization.json.buildJsonObject
+import kotlinx.serialization.json.put
+import org.bouncycastle.asn1.ASN1ObjectIdentifier
+import org.bouncycastle.asn1.x509.ExtendedKeyUsage
+import org.bouncycastle.asn1.x509.Extension
+import org.bouncycastle.asn1.x509.KeyPurposeId
+import org.bouncycastle.asn1.x509.KeyUsage
+import java.security.interfaces.ECPrivateKey
+import java.security.interfaces.ECPublicKey
+import kotlin.random.Random
+
+data class Openid4VpSession(
+    val jwt: String,
+    val privateKey: EcPrivateKey,
+    val nonce: String
+)
+
+fun initiateOpenid4Vp(
+    clientId: String,
+    responseUri: String,
+    state: String
+): Openid4VpSession {
+    val (singleUseReaderKeyPriv, singleUseReaderKeyCertChain) = createSingleUseReaderKey()
+
+    val request = EUPersonalID.getDocumentType().sampleRequests.first { it.id == "full" }
+    val nonce = Random.nextBytes(15).toBase64Url()
+    val publicKey = singleUseReaderKeyPriv.publicKey
+
+    val header = buildJsonObject {
+        put("typ", JsonPrimitive("oauth-authz-req+jwt"))
+        put("alg", JsonPrimitive(publicKey.curve.defaultSigningAlgorithm.jwseAlgorithmIdentifier))
+        put("jwk", publicKey.toJson(null))
+        put("x5c", buildJsonArray {
+            for (cert in singleUseReaderKeyCertChain.certificates) {
+                add(cert.encodedCertificate.toBase64Url())
+            }
+        })
+    }.toString().toByteArray().toBase64Url()
+
+    val body = buildJsonObject {
+        put("client_id", clientId)
+        put("response_uri", responseUri)
+        put("response_type", "vp_token")
+        put("response_mode", "direct_post.jwt")
+        put("nonce", nonce)
+        put("state", state)
+        put("presentation_definition", mdocCalcPresentationDefinition(request))
+        put("client_metadata", calcClientMetadata(singleUseReaderKeyPriv.publicKey))
+    }.toString().toByteArray().toBase64Url()
+
+    val message = "$header.$body"
+    val signature = Crypto.sign(singleUseReaderKeyPriv, Algorithm.ES256, message.toByteArray())
+        .toCoseEncoded().toBase64Url()
+
+    return Openid4VpSession("$message.$signature", singleUseReaderKeyPriv, nonce)
+}
+
+private fun EcPublicKey.toJson(keyId: String?): JsonObject {
+    return JsonWebKey(this).toRawJwk {
+        if (keyId != null) {
+            put("kid", JsonPrimitive(keyId))
+        }
+        put("alg", JsonPrimitive(curve.defaultSigningAlgorithm.jwseAlgorithmIdentifier))
+        put("use", JsonPrimitive("sig"))
+    }
+}
+
+private fun mdocCalcPresentationDefinition(
+    request: DocumentWellKnownRequest
+): JsonObject {
+    return buildJsonObject {
+        // TODO: Fill in a unique ID.
```

Comment:
Any reason we can't make this unique in this PR?

Reply:
I tweaked it. I do not think we rely on it anyway.